griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,145 @@
+import logging
+from schema import Schema, Literal, Optional
+from attr import define, field
+from griptape.artifacts import ErrorArtifact, InfoArtifact, ListArtifact, BlobArtifact
+from griptape.utils.decorators import activity
+from griptape.tools import BaseGoogleClient
+from googleapiclient.errors import HttpError
+from google.auth.exceptions import MalformedError
+from googleapiclient.discovery import Resource
+
+@define
+class GoogleDriveClient(BaseGoogleClient):
+    owner_email: str = field(kw_only=True)
+    LIST_FILES_SCOPES = ['https://www.googleapis.com/auth/drive.readonly']
+    UPLOAD_FILE_SCOPES = ['https://www.googleapis.com/auth/drive.file']
+    max_files: int = field(default=10, kw_only=True)
+
+    def _build_client(self, scopes: list[str]) -> Resource:
+        from google.oauth2 import service_account
+        from googleapiclient.discovery import build
+        credentials = service_account.Credentials.from_service_account_info(
+            self.service_account_credentials, scopes=scopes
+        )
+        delegated_credentials = credentials.with_subject(self.owner_email)
+        service = build('drive', 'v3', credentials=delegated_credentials)
+        return service
+
+    def list_files(self, params: dict) -> ListArtifact | ErrorArtifact:
+        values = params["values"]
+
+        try:
+            service = self._build_client(self.LIST_FILES_SCOPES)
+
+            folder_id = None
+            if "folder_path" in values and values["folder_path"]:
+                folder_id = self._resolve_path_to_id(service, values["folder_path"])
+                if not folder_id:
+                    return ErrorArtifact(f"Could not find folder: {values['folder_path']}")
+
+            if folder_id:
+                query = f"'{folder_id}' in parents and trashed=false"
+            else:
+                query = "mimeType != 'application/vnd.google-apps.folder' and 'root' in parents and trashed=false"
+            max_files = values.get("max_files", self.max_files)
+
+            results = service.files().list(q=query, pageSize=max_files).execute()
+            items = results.get('files', [])
+            return ListArtifact(items)
+
+        except Exception as e:
+            logging.error(e)
+            return ErrorArtifact(f"error retrieving files from Drive: {e}")
+
+    def _resolve_path_to_id(self, service, path: str) -> Optional[str]:
+        parts = path.split('/')
+        current_id = 'root'
+
+        for index, part in enumerate(parts):
+            if index == len(parts) - 1:  # If it's the last part of the path
+                query = f"name='{part}' and '{current_id}' in parents"
+            else:
+                query = f"name='{part}' and '{current_id}' in parents and mimeType='application/vnd.google-apps.folder'"
+
+            response = service.files().list(q=query).execute()
+            files = response.get('files', [])
+
+            if not files:
+                return None
```

Comment:
This will exit the method on any `part` that doesn't return files. Is that what you really want?

Reply:
Yes, I think this is desired behavior for this method given its current use case. if the function returns None, I have put an error message for whichever function that's using the _path_to_file_id.